### PR TITLE
Convert RG_Matrix to GrB_Matrix without writing to internal matrix

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

File: src/graph/rg_matrix/rg_matrix.h [hunk 1]
```diff
@@ -124,11 +124,11 @@ struct _RG_Matrix {
 
 GrB_Info RG_Matrix_new
 (
-    RG_Matrix *A,            // handle of matrix to create
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.h:127]: Why the change in indentation?

File: src/graph/rg_matrix/rg_matrix.c [hunk 2]
```diff
@@ -244,3 +244,59 @@ void RG_Matrix_validateState
 }
 
 
+void RG_Matrix_get
+(
+	RG_Matrix C,
+	GrB_Matrix M
+) {
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.c:251]: I think it would be nicer if we change the signature to:
```
void RG_Matrix_get
(
    GrB_Matrix *M,   // in/output
    RG_Matrix C
) 
```

`M` will allocated within this function, similar to `GrB_Matrix_new`

File: src/graph/rg_matrix/rg_matrix.c [hunk 3]
```diff
@@ -244,3 +244,59 @@ void RG_Matrix_validateState
...
+	ASSERT(C != NULL);
+	ASSERT(A != NULL);
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.c:253]: `ASSERT(M != NULL);`

File: src/graph/rg_matrix/rg_matrix.c [hunk 4]
```diff
@@ -244,3 +244,59 @@ void RG_Matrix_validateState
...
+
+	GrB_Info    info         =  GrB_SUCCESS;
+	GrB_Matrix  m            =  RG_MATRIX_M(C);
+	GrB_Matrix  delta_plus   =  RG_MATRIX_DELTA_PLUS(C);
+	GrB_Matrix  delta_minus  =  RG_MATRIX_DELTA_MINUS(C);
+
+	info = GrB_Matrix_dup(&M, m);
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.c:260]: Consider using `GrB_transpose`, given that the output matrix been allocated.
```
GrB_transpose(M, dm, NULL, m, GrB_DESC_RSCT0);
```
As `T(T(m)) = m`
this will perform both duplication and masking at once
Also consider renaming `M` to `A` as this is now confusing

Need to check which of the two methods is faster

File: src/graph/rg_matrix/rg_matrix.h [hunk 5]
```diff
@@ -304,3 +304,10 @@ void RG_Matrix_free
 	RG_Matrix *C
 );
 
+// get matrix C without writing to internal matrix
+void RG_Matrix_get
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.h:308]: I think we need to find a better name for this method, also please add test for it

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/graph/rg_matrix/rg_matrix.h (hunk 1 above)
Comment (AviAvni) [reply src/graph/rg_matrix/rg_matrix.h:127]: From spaces to tabs as the format we use

File: src/graph/rg_matrix/rg_matrix.h (hunk 5 above)
Comment (AviAvni) [reply src/graph/rg_matrix/rg_matrix.h:308]: Maybe RG_Matrix_export?

File: tests/unit/test_rg_matrix.cpp [hunk 6]
```diff
@@ -737,3 +737,173 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
 	free(I);
 	free(J);
 }
+
+
+TEST_F(RGMatrixTest, RGMatrix_export_no_changes) {
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:742]: Add a comment explaining the purpose of this test, and in general what's going on here, (introduction)

File: tests/unit/test_rg_matrix.cpp [hunk 7]
```diff
@@ -737,3 +737,173 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
+	GrB_Type    t                   =  GrB_BOOL;
+	RG_Matrix   A                   =  NULL;
+	GrB_Matrix  M                   =  NULL;
+	GrB_Matrix  N                   =  NULL;
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:746]: Please add a comment for each variable you see fit, explaining its purpose

File: tests/unit/test_rg_matrix.cpp [hunk 8]
```diff
@@ -737,3 +737,173 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
+	ASSERT_EQ(info, GrB_SUCCESS);
+
+	// set element at position i,j
+	info = RG_Matrix_setElement_BOOL(A, true, i, j);
+	ASSERT_EQ(info, GrB_SUCCESS);
+
+	// get internal matrices
+	M   =  RG_MATRIX_M(A);
+	DP  =  RG_MATRIX_DELTA_PLUS(A);
+	DM  =  RG_MATRIX_DELTA_MINUS(A);
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:769]: Unused, please remove

File: tests/unit/test_rg_matrix.cpp [hunk 9]
```diff
@@ -737,3 +737,173 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
+	ASSERT_EQ(nvals, 1);
+
+	// M should be empty
+	M_NOT_EMPTY();
+
+	// DM should be empty
+	DM_EMPTY();
+
+	// DP should contain a single element
+	DP_EMPTY();
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:789]: Please remove unnecessary in this test context

File: tests/unit/test_rg_matrix.cpp [hunk 10]
```diff
@@ -737,3 +737,173 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
+	//--------------------------------------------------------------------------
+
+	info = RG_Matrix_export(&N, A);
+	ASSERT_EQ(info, GrB_SUCCESS);
+
+	//--------------------------------------------------------------------------
+	// validation
+	//--------------------------------------------------------------------------
+
+	Compare_Matrices(M, N, t, nrows, ncols);
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:802]: Please change to: `ASSERT_GrB_Matrices_EQ(M, N);` ?

File: tests/unit/test_rg_matrix.cpp [hunk 11]
```diff
@@ -737,3 +737,173 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
+
+	// clean up
+	RG_Matrix_free(&A);
+	ASSERT_TRUE(A == NULL);
+	GrB_Matrix_free(&N);
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:807]:
```
	GrB_Matrix_free(&N);
	RG_Matrix_free(&A);
	ASSERT_TRUE(A == NULL);
```

File: tests/unit/test_rg_matrix.cpp [hunk 12]
```diff
@@ -737,3 +737,173 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
+}
+
+TEST_F(RGMatrixTest, RGMatrix_export_pending_changes) {
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:810]: Apply the same comments from the previous function here

File: src/graph/rg_matrix/rg_matrix.h [hunk 13]
```diff
@@ -304,3 +304,10 @@ void RG_Matrix_free
 	RG_Matrix *C
 );
 
+// get matrix C without writing to internal matrix
+GrB_Info RG_Matrix_export
+(
+	GrB_Matrix *A,
+	RG_Matrix C
+);
+
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.h:313]: `Free` is always the last function in a head file,
please relocate.

File: src/graph/rg_matrix/rg_matrix.c [hunk 14]
```diff
@@ -244,3 +244,75 @@ void RG_Matrix_validateState
...
+GrB_Info RG_Matrix_export
+(
+	GrB_Matrix *A,
+	RG_Matrix C
+) {
+	ASSERT(C != NULL);
+	ASSERT(A != NULL);
+
+	GrB_Info    info         =  GrB_SUCCESS;
+	GrB_Matrix  a            =  rm_calloc(1, sizeof(GrB_Matrix));
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.c:256]: Init `a` to NULL

File: src/graph/rg_matrix/rg_matrix.c [hunk 15]
```diff
@@ -244,3 +244,75 @@ void RG_Matrix_validateState
...
+	bool  additions  =  delta_plus_nvals  >  0;
+	bool  deletions  =  delta_minus_nvals  >  0;
+
+	//--------------------------------------------------------------------------
+	// perform copy and deletions if needed
+	//--------------------------------------------------------------------------
+	
+	// in case there are items to delete use mask otherwise just copy
+	GrB_Matrix mask = deletions ? delta_minus : NULL;
+	GrB_Descriptor desc = deletions ? GrB_DESC_RSCT0 : GrB_DESC_RST0;
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.c:292]: in case there are no deletions, use desc as  `GrB_DESC_RT0`

File: src/graph/rg_matrix/rg_matrix.c [hunk 16]
```diff
@@ -244,3 +244,75 @@ void RG_Matrix_validateState
...
+	info = GrB_transpose(a, mask, NULL, m, desc);
+	ASSERT(info == GrB_SUCCESS);
+
+	info = GrB_wait(&delta_plus);
+	ASSERT(info == GrB_SUCCESS);
+
+	info = GrB_wait(&delta_minus);
+	ASSERT(info == GrB_SUCCESS);
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.c:300]: move upwards, before calling `nvals` on either `dp` or `dm`

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/unit/test_rg_matrix.cpp [hunk 17]
```diff
@@ -598,7 +598,7 @@ TEST_F(RGMatrixTest, RGMatrix_managed_transposed) {
 }
 
 // nvals(A eWiseMult B) == nvals(A) == nvals(B)
-void Compare_Matrices(GrB_Matrix m, GrB_Matrix n, GrB_Type t, GrB_Index nrows, GrB_Index ncols)
+void ASSERT_GrB_Matrices_EQ(GrB_Matrix m, GrB_Matrix n, GrB_Type t, GrB_Index nrows, GrB_Index ncols)
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:601]: Please change sig to: `void ASSERT_GrB_Matrices_EQ(const GrB_Matrix A, const  GrB_Matrix B)`

File: tests/unit/test_rg_matrix.cpp [hunk 18]
```diff
@@ -737,3 +737,141 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
+	//--------------------------------------------------------------------------
+	// flush matrix, sync
+	//--------------------------------------------------------------------------
+	
+	// wait, force sync
+	sync = true;
+	RG_Matrix_wait(A, sync);
+
+	RG_Matrix_nvals(&nvals, A);
+	ASSERT_EQ(nvals, 1);
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:777]: Please remove

File: tests/unit/test_rg_matrix.cpp [hunk 19]
```diff
@@ -737,3 +737,141 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
+
+	info = RG_Matrix_new(&A, t, nrows, ncols, multi_edge, maintain_transpose);
+	ASSERT_EQ(info, GrB_SUCCESS);
+
+	// set element at position i,j
+	info = RG_Matrix_setElement_BOOL(A, true, i, j);
+	ASSERT_EQ(info, GrB_SUCCESS);
+
+	// get internal matrices
+	M   =  RG_MATRIX_M(A);
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:766]: Remove extra space between `M` and `=` `M   =  RG_MATRIX_M(A);`

File: tests/unit/test_rg_matrix.cpp [hunk 20]
```diff
@@ -737,3 +737,141 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:824]: remove extra spece

File: tests/unit/test_rg_matrix.cpp [hunk 21]
```diff
@@ -737,3 +737,141 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
+	//--------------------------------------------------------------------------
+	// flush matrix, sync
+	//--------------------------------------------------------------------------
+	
+	// wait, force sync
+	sync = true;
+	RG_Matrix_wait(A, sync);
+
+	RG_Matrix_nvals(&nvals, A);
+	ASSERT_EQ(nvals, 1);
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:835]: Remove

File: tests/unit/test_rg_matrix.cpp [hunk 22]
```diff
@@ -737,3 +737,141 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
...
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:865]: Remove

File: src/graph/rg_matrix/rg_matrix.c [hunk 23]
```diff
@@ -244,3 +244,74 @@ void RG_Matrix_validateState
...
+	ASSERT(A != NULL);
+
+	GrB_Info    info         =  GrB_SUCCESS;
+	GrB_Matrix  a            =  NULL;
+	GrB_Matrix  m            =  RG_MATRIX_M(C);
+	GrB_Matrix  delta_plus   =  RG_MATRIX_DELTA_PLUS(C);
+	GrB_Matrix  delta_minus  =  RG_MATRIX_DELTA_MINUS(C);
+	GrB_Type    t;
+	GrB_Index   nrows;
+	GrB_Index   ncols;
```
Comment (swilly22) [inline src/graph/rg_matrix/rg_matrix.c:262]:
```
	GrB_Type    t;
	GrB_Index   ncols;
	GrB_Index   nrows;
	GrB_Matrix  a                     =  NULL;
	GrB_Matrix  m                    =  RG_MATRIX_M(C);
	GrB_Info    info                   =  GrB_SUCCESS;
	GrB_Matrix  delta_plus     =  RG_MATRIX_DELTA_PLUS(C);
	GrB_Matrix  delta_minus  =  RG_MATRIX_DELTA_MINUS(C);
```

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/unit/test_rg_matrix.cpp [hunk 24]
```diff
@@ -720,12 +732,12 @@ TEST_F(RGMatrixTest, RGMatrix_fuzzy) {
 	// validation
 	//--------------------------------------------------------------------------
 
-	ASSERT_GrB_Matrices_EQ(M, N, t, nrows, ncols);
+	ASSERT_GrB_Matrices_EQ(M, N);
 
 	info = GrB_transpose(NT, NULL, NULL, N, NULL);
```
Comment (swilly22) [inline tests/unit/test_rg_matrix.cpp:737]: `GrB_Matrix_new(&NT, ...)`
Don't forget clean-up

**Review by swilly22 — CHANGES_REQUESTED** [review]
